includes/services/extended/instagram-basic-display.php

```diff
@@ -0,0 +1,193 @@
+		$params['body']['redirect_uri'] = $this->redirect_uri;
+		return $params;
+	}
+
+	/**
+	 * Add scope to the outbound URL, and allow developers to modify it, and also
+	 * pack all of the redirct_uri params into the state param as Instagram strips
+	 * these before redirecting
+	 * @param  array $params Core request parameters
+	 * @return Array containing originals, plus the scope parameter, and a serialized state param
+	 */
+	function filter_request_token( $params ) {
+		$params['scope'] = apply_filters( 'keyring_' . $this->get_name() . '_scope', 'user_profile,user_media' );
+
+		// The Instagram API does not return redirect_uri params, so we need to pack these
+		// into the state param
+		$url_components = parse_url( $params['redirect_uri'] );
+		parse_str( $url_components['query'], $redirect_state );
+		$redirect_state['state'] = $params['state'];
+		$params['state']         = Keyring_Util::get_hashed_parameters( $redirect_state );
+		$params['redirect_uri']  = $this->redirect_uri;
+
+		return $params;
+	}
+
+	function basic_ui_intro() {
+		/* translators: url */
+		echo '<p>' . sprintf( __( 'To get started, <a href="%1$s">register an OAuth client on Instagram</a>. The most important setting is the <strong>OAuth redirect_uri</strong>, which should be set to <code>%2$s</code>. You can set the other values to whatever you like.', 'keyring' ), 'http://instagram.com/developer/clients/register/', Keyring_Util::admin_url( $this->get_name(), array( 'action' => 'verify' ) ) ) . '</p>';
+		echo '<p>' . __( "Once you've saved those changes, copy the <strong>CLIENT ID</strong> value into the <strong>API Key</strong> field, and the <strong>CLIENT SECRET</strong> value into the <strong>API Secret</strong> field and click save (you don't need an App ID value for Instagram).", 'keyring' ) . '</p>';
+	}
+
+	function build_token_meta( $token ) {
+		$this->set_token(
+			new Keyring_Access_Token(
+				$this->get_name(),
+				$token['access_token'],
+				array()
+			)
+		);
+		$response = $this->request( add_query_arg( 'fields', 'id, username', $this->self_url ), array( 'method' => $this->self_method ) );
+
+		if ( Keyring_Util::is_error( $response ) ) {
+			$meta = array();
+		} else {
+			$meta = array(
+				'user_id' => $response->id,
+				'name'    => $response->username,
+			);
+			if ( ! empty( $token['expires_in'] ) ) {
+				$meta['expires'] = time() + $token['expires_in'];
+			}
+		}
+
+		return apply_filters( 'keyring_access_token_meta', $meta, $this->get_name(), $token, null, $this );
+	}
+
+	function get_display( Keyring_Access_Token $token ) {
+		return $token->get_meta( 'name' );
+	}
+
+	function test_connection() {
+		$response = $this->request( $this->self_url, array( 'method' => $this->self_method ) );
+		if ( ! Keyring_Util::is_error( $response ) ) {
+			return true;
+		}
+
+		return $response;
+	}
+
+	function parse_access_token( $token ) {
+		//Somewhat of a hack, but we need a place to exchange for a long lived token
```

Comment:
I can't find a better place to do this. We're using this as a way to exchange the token at the right place in the flow, but it doesn't seem the most elegant solution. 

Perhaps in the future, we could add an additional hook for this purpose

Reply:
I've noticed that we should be doing [this long-lived token exchange ](https://developers.facebook.com/docs/facebook-login/access-tokens/refreshing)for Facebook too. 

Another option could be to create a `0-facebook-base.php` service, like we have for the Google services, and handle this along with the deletion flows from within there.

Let's leave this here for now (unless we can think of somewhere better) and create an issue so we can revisit it.